wayland-backend/src/rs/client_impl/mod.rs

```diff
@@ -772,3 +571,224 @@ impl InnerHandle {
+                            user_data: child_udata,
+                            serial: guard.next_serial(),
+                        }
+                    };
+
+                    let child_id = InnerObjectId { id: new_id, serial: child_obj.data.serial, interface: child_obj.interface };
+                    created_id = Some(child_id.clone());
+
+                    if let Err(()) = guard.map.insert_at(new_id, child_obj) {
+                        // abort parsing, this is an unrecoverable error
+                        let err = WaylandError::Protocol(ProtocolError {
+                            code: 0,
+                            object_id: 0,
+                            object_interface: "".into(),
+                            message: format!(
+                                "Protocol error: server tried to create \
+                                an object \"{}\" with invalid id {}.",
+                                child_interface.name, new_id
+                            ),
+                        });
+                        return Err(guard.store_and_return_error(err));
+                    }
+
+                    Argument::NewId(ObjectId { id: child_id })
+                }
+            });
+        }
+
+        if guard.debug {
+            super::debug::print_dispatched_message(
+                receiver.interface.name,
+                message.sender_id,
+                message_desc.name,
+                &args,
+            );
+        }
+
+        // If this event is send to an already destroyed object (by the client), swallow it
+        if receiver.data.client_destroyed {
+            // but close any associated FD to avoid leaking them
+            for a in args {
+                if let Argument::Fd(fd) = a {
+                    let _ = ::nix::unistd::close(fd);
+                }
+            }
+            continue;
+        }
+
+        // Invoke the user callback
+        let id = InnerObjectId {
+            id: message.sender_id,
+            serial: receiver.data.serial,
+            interface: receiver.interface,
+        };
+
+        // unlock the mutex while we invoke the user callback
+        std::mem::drop(guard);
+        log::debug!("Dispatching {}.{} ({})", id, receiver.version, DisplaySlice(&args));
+        let ret = receiver
+            .data
+            .user_data
+            .clone()
+            .event(&backend, Message { sender_id: ObjectId { id }, opcode: message.opcode, args });
+        // lock it again to resume dispatching
+        guard = backend.backend.state.lock_protocol();
+
+        // If this event is a destructor, destroy the object
+        if message_desc.is_destructor {
+            guard
+                .map
+                .with(message.sender_id, |obj| {
+                    obj.data.server_destroyed = true;
+                    obj.data.client_destroyed = true;
+                })
+                .unwrap();
+            receiver.data.user_data.destroyed(ObjectId {
```

Comment:
This may dispatch, causing a deadlock if the user interfaces with the backend again?

Reply:
If you mean if the user tries to start a dispatch from within a callback, yeah, that would deadlock. You're not meant to do that.

Reply:
Maybe not dispatch but send a request?

Reply:
Obviously not to the same object, but a different one